test/jit/test_union.py

```diff
@@ -0,0 +1,345 @@
+                    r" Union\[str, int\] but instead found type "
+                    r"List\[str\]"):
+            scripted(["foo", "bar", "baz"])
+
+    def test_union_return_type(self):
+        def fn(x: int) -> Union[int, str]:
+            return "foo"
+
+        self.checkScript(fn, (1,))
+
+    def test_union_return_type_with_same_type_branching(self):
+        def fn(x: int) -> Union[int, str]:
+            if x % 2:
+                return "foo"
+            else:
+                return "bar"
+
+        self.checkScript(fn, (1,))
+        self.checkScript(fn, (8,))
+
+    def test_union_return_type_with_different_type_branching(self):
+        def fn(x: int) -> Union[int, str]:
+            if x % 2:
+                return "foo"
+            else:
+                return 1
+
+        self.checkScript(fn, (1,))
+        self.checkScript(fn, (8,))
+
+    def test_union_as_annotation(self):
+        def fn() -> Union[int, str]:
+            x: Union[int, str] = "foo"
+            return x
+
+        self.checkScript(fn, ())
+
+    def test_union_as_annotation_in_typed_container(self):
+        def fn() -> None:
+            l: List[Union[int, str]] = []
+            u1: Union[int, str] = "foo"
+            u2: Union[int, str] = 1
+            l.append(u1)
+            l.append(u2)
+
+        self.checkScript(fn, ())
+
+    def test_union_as_internal_tuple_type(self):
+        def fn():
+            t: Tuple[Union[int, str], Union[int, str]] = (1, "foo")
+            return t
+
+        self.checkScript(fn, ())
+
+    def test_union_variable_can_be_reassigned(self):
+        def fn() -> Union[int, str]:
+            x: Union[int, str] = "foo"
+            i: int = 1
+            s: str = "bar"
+            x = i
+            x = s
+            return x
+
+        self.checkScript(fn, ())
+
+    def test_union_does_not_replace_existing_annotated_type(self):
+        def fn():
+            x: List[int] = [1, 2, 3]
+            x.append("foo")
+            return x
+
+        with self.assertRaisesRegex(RuntimeError, "Could not match type str"):
+            scripted = torch.jit.script(fn)
+            scripted()
+
+    def test_union_does_not_replace_existing_union_annotated_type(self):
+        def fn():
+            x: List[Union[int, str]] = [1, "foo", 3]
+            x.append(2.0)
+            return x
+
+        with self.assertRaisesRegex(RuntimeError, "Could not match type float"):
+            scripted = torch.jit.script(fn)
+            scripted()
+
+    def test_union_does_not_replace_existing_annotated_type_with_empty_container(self):
+        def fn():
+            x: List[int] = []
+            x.append("foo")
+            return x
+
+        with self.assertRaisesRegex(RuntimeError, "Could not match type str"):
+            scripted = torch.jit.script(fn)
+            scripted()
+
+    def test_unions_of_unions_are_flattened(self):
+        @torch.jit.script
+        def fn_with_union(x: Union[int, str, float]) -> str:
```

Comment:
This `fn_with_union` doesn't seem necessary, as long as TorchScript accepts an `int` into `fn_with_nested_union` the test can prove union is flattened.

Same for a few other test cases where two functions are scripted, in those cases the first function doesn't seem necessary, for example `test_unions_of_a_single_argument_vanish `

Optional:
use `FileCheck` to verify union is flattened